### PR TITLE
Code review and cleanup

### PR DESCRIPTION
**Pull request by dkarrasch**
This is an unsolicited :) review and cleanup of the LinearMap part. One major motivation was to add type parameters for the specific types of matrix fields, in order to have concrete types and make the compiler aware of the types of the struct fields. This follows a recommendation from the Julia performance tipps. Also, I tried to reduce the allocation of intermediate (mostly) vectors. Other changes are more of style nature, e.g., I noticed that the indentation was quite irregular, so I fixed that along the way. Feel free to ask for partial reversions or changes as you prefer.

Do you have some benchmarks to see if there are any improvements or regressions?

### CODE REVIEW

File: src/MatrixEquations.jl [hunk 1]
```diff
@@ -19,7 +20,7 @@ export lanv2, ladiv, lag2, lacn2!
 export lyapc, lyapd, lyapcs!, lyapds! 
 export plyapc, plyaps, plyapcs!, plyapd, plyapds!  
 export arec, ared, garec, gared
-export sylvc, sylvd, gsylv, sylvcs!, sylvds!, gsylvs!, gsylvs1!
```
Comment (dkarrasch) [inline src/MatrixEquations.jl:22]: This didn't exist anywhere in the package.

File: src/meoperators.jl [hunk 2]
```diff
@@ -1,184 +1,174 @@
...
-lyapop(A::Schur{T,Matrix{T}}; kwargs...) where T = LyapunovMap(A.T; kwargs...)
-lyapop(A::Number; kwargs...) = LyapunovMap(A*ones(eltype(A),1,1); kwargs...) 
+lyapop(A::Schur; kwargs...) = LyapunovMap(A.T; kwargs...)
+lyapop(A::Number; kwargs...) = LyapunovMap(fill(A, 1, 1); kwargs...)
 function Base.size(L::LyapunovMap)
-    n = size(L.A,1)
-    N = L.her ? Int(n*(n+1)/2) : n*n
-    return (N,N)
-end
-Base.size(L::Adjoint{<:Any, <: LyapunovMap}) = size(L.parent)
```
Comment (dkarrasch) [inline src/meoperators.jl:73]: Such methods are covered by the generic `size` method in `LinearMaps.jl`.

File: src/meoperators.jl [hunk 3]
```diff
@@ -188,684 +178,658 @@ For the definitions of the Lyapunov operators see:
 M. Konstantinov, V. Mehrmann, P. Petkov. On properties of Sylvester and Lyapunov
 operators. Linear Algebra and its Applications 312:35–71, 2000.
 """
-function lyapop(A::AbstractMatrix{T}, E::AbstractMatrix{T}; disc = false, her = false) where T
-    GeneralizedLyapunovMap(A, E; disc = disc, her = her)
+function lyapop(A::AbstractMatrix{T1}, E::AbstractMatrix{T2}; disc=false, her=false) where {T1,T2}
+   T = promote_type(T1, T2)
+   GeneralizedLyapunovMap(convert.(AbstractMatrix{T}, (A, E))...; disc=disc, her=her)
```
Comment (dkarrasch) [inline src/meoperators.jl:183]: This conversion is a no-op if both types are equal, so there is no need for another "homogeneous-type" method.

File: src/meoperators.jl [hunk 4]
```diff
@@ -188,684 +178,658 @@ For the definitions of the Lyapunov operators see:
...
 end
-Base.size(L::Adjoint{<:Any, <: GeneralizedLyapunovMap}) = size(L.parent)
-function LinearAlgebra.mul!(y::AbstractVector, L::GeneralizedLyapunovMap{T}, x::AbstractVector) where T
-   n = size(L.A,1)
+function mul!(y::AbstractVector, L::GeneralizedLyapunovMap{T}, x::AbstractVector) where T
+   n = size(L.A, 1)
    T1 = promote_type(T, eltype(x))
    if L.her
-      T == T1 ?  X = vec2triu(x, her = true) :  X = vec2triu(convert(Vector{T1}, x), her = true)
+      X = vec2triu(T == T1 ? x : convert(Vector{T1}, x), her=true)
```
Comment (dkarrasch) [inline src/meoperators.jl:199]: I wonder why you do all the conversion stuff manually. What would happen if `X` was real and `L` complex?

**Review by dkarrasch — COMMENTED** [review]
A few comments to help the review.

File: src/MatrixEquations.jl (hunk 1 above)
Comment (dkarrasch) [reply src/MatrixEquations.jl:22]: I mean `gsylvs1!`.

Comment (andreasvarga) [reply src/MatrixEquations.jl:22]: It's OK. Probably it is reminiscent from a test version. Thanks.

**Comment by andreasvarga** [comment]
I merged and corrected later some small errors. I think ME is now ready for a version bump.

**Comment by dkarrasch** [comment]
As you like. I started with the `Adjoint` stuff, but I think it will be only non-user-facing changes, unless people rely on your new `*Map`-types. But since these are new, users probably won't.

**Pull request merged**